kernel/posix/pthread.c

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <atomic.h>
+#include "ksched.h"
+#include "wait_q.h"
+
+/* Kernel Structure for Number of threads */
+struct k_thread threads[CONFIG_MAX_PTHREAD_COUNT];
+
+static int is_prio_valid(int priority)
+{
+	if ((priority < K_HIGHEST_APPLICATION_THREAD_PRIO)
+	    || (priority > K_HIGHEST_APPLICATION_THREAD_PRIO)) {
+		return -1;
+	} else       {
+		return 0;
+	}
+
+}
+
+int pthread_attr_setschedparam(pthread_attr_t *attr,
+			       const struct sched_param *schedparam)
+{
+	int priority = schedparam->priority;
+
+	/*
+	 *  Task priorities can range from  K_LOWEST_THREAD_PRIO to
+	 *  K_HIGHEST_THREAD_PRIO.
+	 */
+	if (is_prio_valid(priority) != 0) {
+		/* Invalid Prioritty */
+		return EINVAL;
+	}
+	attr->priority = priority;
+	return 0;
+}
+
+int pthread_attr_setstack(pthread_attr_t *attr, void *stackaddr,
+			  size_t stacksize)
+{
+
+	if ((stackaddr == NULL) || (stacksize <= K_LOWEST_STACK_SIZE)) {
+		return EINVAL;
+	}
+	attr->stack = stackaddr;
+	attr->stacksize = stacksize;
+	return 0;
+}
+
+static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
+{
+	void * (*fun_ptr)(void *);
+
+	fun_ptr = arg3;
+	(*fun_ptr)((void *)arg1);
+}
+
+int pthread_create(pthread_t *newthread, const pthread_attr_t *attr,
+		   void *(*threadroutine)(void *), void *arg)
+{
+	static int thread_count;
+
+	if ((attr == NULL) || (attr->stack == NULL)) {
+		return EINVAL;
+	}
+	if (thread_count < CONFIG_MAX_PTHREAD_COUNT) {
+		*newthread = (pthread_t) k_thread_create(&threads[thread_count],
+			attr->stack, attr->stacksize,
+			(k_thread_entry_t)zephyr_thread_wrapper, (void *)arg,
+			NULL, threadroutine, attr->priority, attr->options,
+			attr->delayedstart);
+		atomic_inc(&thread_count);
```

Comment:
Does this really work in all cases?
Seeme to me that two competing threads could get inside this IF condition with the same value of thread_count, and then try to run k_thread_create() on the same index.
It would be better to have some kind of test-and-increment going on in the conditional, so that no two threads can enter the if block with the same value of thread_count.

Reply:
There is a race condition here; there are comments about it in another pull request.

Reply:
*bangs head on wall*